sys/fido2/ctap/transport/hid/ctap_hid.c

```diff
@@ -0,0 +1,711 @@
+void usb_hid_io_init(usbus_t *usbus, const uint8_t *report_desc,
+                     size_t report_desc_size);
+
+/**
+ * @brief Indicate if authenticator is busy processing a transactions
+ *
+ * Transactions are atomic, therefore only 1 transaction can be processed at
+ * once
+ */
+static bool g_is_busy = false;
+
+/**
+ * @brief State for handling transactions
+ */
+static ctap_hid_state_t g_state;
+
+/**
+ * @brief Logical CTAPHID channels
+ */
+static ctap_hid_cid_t g_cids[CTAP_HID_CIDS_MAX];
+
+/**
+ * @brief Incremental channel ids
+ *
+ * channel id 0 is reserved
+ */
+static uint32_t g_cid = 1;
+
+void fido2_ctap_transport_hid_create(void)
+{
+    usbdev_t *usbdev = usbdev_get_ctx(0);
+
+    assert(usbdev);
+    usbus_init(&g_usbus, usbdev);
+    usb_hid_io_init(&g_usbus, report_desc_ctap, sizeof(report_desc_ctap));
+
+    usbus_create(g_usb_stack, USBUS_STACKSIZE, USBUS_PRIO, USBUS_TNAME, &g_usbus);
+}
+
+void fido2_ctap_transport_hid_handle_packet(void *pkt_raw)
+{
+    ctap_hid_pkt_t *pkt = (ctap_hid_pkt_t *)pkt_raw;
+    uint32_t cid = pkt->cid;
+    uint8_t status = CTAP_HID_BUFFER_STATUS_BUFFERING;
+
+    if (cid == 0x00) {
+        /* cid = 0x00 always invalid */
+        send_error_response(cid, CTAP_HID_ERR_INVALID_CHANNEL);
+        return;
+    }
+    else if (g_is_busy) {
+        if (g_state.cid == cid) {
+            /* CTAP specification (version 20190130) section 8.1.5.3 */
+            if (is_init_type_pkt(pkt)) {
+                if (pkt->init.cmd == CTAP_HID_COMMAND_INIT) {
+                    /* abort */
+                    clear_ctap_buffer();
+                    status = buffer_pkt(pkt);
+                }
+                else if (g_state.is_locked && pkt->init.cmd ==
+                         CTAP_HID_COMMAND_CANCEL) {
+                    g_state.should_cancel = true;
+                }
+                /* random init type pkt. invalid sequence of pkts */
+                else {
+                    send_error_response(cid, CTAP_HID_ERR_INVALID_SEQ);
+                    return;
+                }
+            }
+            /* packet for this cid is currently being worked */
+            else if (g_state.is_locked) {
+                send_error_response(cid, CTAP_HID_ERR_CHANNEL_BUSY);
+                return;
+            }
+            else {
+                /* buffer cont packets */
+                status = buffer_pkt(pkt);
+            }
+        }
+        /* transactions are atomic. Deny all other cids if busy with one cid */
+        else {
+            send_error_response(cid, CTAP_HID_ERR_CHANNEL_BUSY);
+            return;
+        }
+    }
+    else {
+        /* first init packet received starts a transaction */
+        if (is_init_type_pkt(pkt)) {
+            g_is_busy = true;
+            status = buffer_pkt(pkt);
+        }
+        /* ignore rest */
+    }
+
+    if (status == CTAP_HID_BUFFER_STATUS_ERROR) {
+        send_error_response(cid, g_state.err);
```

Comment:
Referring to [this comment](https://github.com/RIOT-OS/RIOT/pull/16489#discussion_r676511702): This one does not miss a `return`?

Reply:
Since the function exists anyways after this If statement there is no need for a return. I could insert one though if it makes the code more readable ?

Reply:
Ah I see. Well, I think I would add it anyway.